File: LeetCode/Easy/ValidParentheses.cpp
class Solution {
public:
    bool matching(char a,char b){
        return ((a=='(' && b==')')
        || (a=='[' && b==']')
        ||(a=='{' && b=='}'));
    }
    bool isValid(string str) {
        stack<int> s;
        for(int i=0;i<str.length();i++){
            if(str[i] == '(' || str[i] == '[' || str[i] =='{')
                s.push(str[i]);
            else{
                if(s.empty())
                    return false;
                else if(!matching(s.top(),str[i]))
                    return false;
                else
                    s.pop();
            }
        }
        return (s.empty()==true);
    }
};

File: MAANG/ValidParentheses.java
class Solution {
    boolean compare(char a, char b) {
        return (a == '(' && b == ')' ||
                a == '{' && b == '}' ||
                a == '[' && b == ']');
    }
    public boolean isValid(String s) {
        Deque<Character> stack = new ArrayDeque<Character>();
        for(int i=0;i<s.length();i++){
            if(s.charAt(i) == '(' || s.charAt(i) =='[' || s.charAt(i) =='{')
                stack.push(s.charAt(i));
            else{
                if(stack.isEmpty())
                    return false;
                else if(!compare(stack.peek(),s.charAt(i)))
                    return false;
                else
                    stack.pop();
            }
        }
        return stack.isEmpty();
    }
}
